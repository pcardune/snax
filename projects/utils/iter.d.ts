interface PeakableIterator<T> extends Iterator<T> {
    peek(): IteratorResult<T>;
}
declare class Peakable<T> implements PeakableIterator<T> {
    private iterator;
    private buffer;
    constructor(iterator: Iterator<T>);
    next(): IteratorResult<T>;
    peek(): IteratorResult<T>;
}
export declare function peakable<T>(it: Iterator<T>): Peakable<T>;
export declare function collect<T>(it: Iterator<T, T>): T[];
export declare function map<I, O>(it: Iterator<I>, map: (i: I) => O): IterableIterator<O>;
declare class CharCodeIterator implements IterableIterator<number> {
    private input;
    private index;
    constructor(input: string);
    next(): IteratorResult<number>;
    prefix(): string;
    suffix(): string;
    [Symbol.iterator](): this;
}
export declare function charCodes(input: string): CharCodeIterator;
declare class ConcatIterator<T> implements Iterator<T> {
    private iters;
    private index;
    constructor(...iters: Iterator<T>[]);
    next(): IteratorResult<T>;
}
export declare function concat<T>(...iters: Iterator<T>[]): ConcatIterator<T>;
export declare class RewindableIterator<T> implements IterableIterator<T> {
    private iter;
    private buffer;
    private index;
    constructor(iter: Iterator<T>);
    next(): IteratorResult<T>;
    reset(n: number): void;
    [Symbol.iterator](): this;
    get buffered(): number;
}
export declare class BacktrackableIterator<T> implements IterableIterator<T> {
    private iter;
    private buffer;
    constructor(iter: Iterator<T>);
    [Symbol.iterator](): this;
    next(): IteratorResult<T>;
    pushBack(item: T): void;
}
export declare function backtrackable<T>(iter: Iterator<T>): BacktrackableIterator<T>;
export declare function rewindable<T>(iter: Iterator<T>): RewindableIterator<T>;
export declare function iterable<T>(it: Iterator<T>): IterableIterator<T>;
export declare function range(start: number, end: number): Generator<number, void, unknown>;
export {};
//# sourceMappingURL=iter.d.ts.map