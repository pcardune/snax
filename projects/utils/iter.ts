export abstract class Iter<T> implements IterableIterator<T> {
  [Symbol.iterator]() {
    return this;
  }

  abstract next(): IteratorResult<T>;

  map<O>(mapper: (i: T) => O): Iter<O> {
    return new MapIter(this, mapper);
  }

  chain(...iters: Iter<T>[]): Iter<T> {
    return new ChainIter(...iters);
  }

  first(): T | undefined {
    let next = this.next();
    return next.value;
  }

  toArray(): T[] {
    return [...this];
  }

  join(sep: string): string {
    let s = '';
    let next = this.next();
    if (next.done) {
      return s;
    }
    for (const item of this) {
      s += sep;
      s += item;
    }
    return s;
  }
}

class PlainIter<T> extends Iter<T> {
  private iterator: Iterator<T>;
  constructor(iterable: Iterable<T>) {
    super();
    this.iterator = iterable[Symbol.iterator]();
  }
  next() {
    return this.iterator.next();
  }
}

export function iter<T>(iterable: Iterable<T> = []) {
  return new PlainIter(iterable);
}

class ChainIter<T> extends Iter<T> {
  private iters: Iter<T>[];
  constructor(...iters: Iter<T>[]) {
    super();
    this.iters = iters;
  }
  next(): IteratorResult<T, undefined> {
    while (this.iters.length > 0) {
      const next = this.iters[0].next();
      if (next.done) {
        this.iters.shift();
      } else {
        return next;
      }
    }
    return { done: true, value: undefined };
  }
}

class MapIter<I, O> extends Iter<O> {
  private iterator: Iterator<I>;
  private mapper: (i: I) => O;
  constructor(iterable: Iterable<I>, mapper: (i: I) => O) {
    super();
    this.iterator = iterable[Symbol.iterator]();
    this.mapper = mapper;
  }
  next(): IteratorResult<O> {
    const { value, done } = this.iterator.next();
    if (done) {
      return { done, value: undefined };
    }
    return { value: this.mapper(value), done: false };
  }
}

interface PeakableIterator<T> extends Iterator<T> {
  peek(): IteratorResult<T>;
}

class Peakable<T> implements PeakableIterator<T> {
  private iterator: Iterator<T>;
  private buffer: IteratorResult<T> | null;
  constructor(iterator: Iterator<T>) {
    this.iterator = iterator;
    this.buffer = null;
  }
  next(): IteratorResult<T> {
    let result = this.peek();
    this.buffer = null;
    return result;
  }
  peek(): IteratorResult<T> {
    if (this.buffer == null) {
      this.buffer = this.iterator.next();
      return this.buffer;
    }
    return this.buffer;
  }
}

export function peakable<T>(it: Iterator<T>) {
  return new Peakable(it);
}

export function collect<T>(it: Iterator<T, T>): T[] {
  let values: T[] = [];
  let result;
  do {
    result = it.next();
    if (result.done == false) {
      values.push(result.value);
    }
  } while (result.done == false);
  return values;
}

export function map<I, O>(
  it: Iterable<I>,
  map: (i: I) => O
): IterableIterator<O> {
  return new MapIter(it, map);
}

class CharCodeIterator implements IterableIterator<number> {
  private input: string;
  private index: number = 0;
  constructor(input: string) {
    this.input = input;
  }
  next(): IteratorResult<number> {
    if (this.index >= this.input.length) {
      return { done: true, value: undefined };
    }
    return { done: false, value: this.input.charCodeAt(this.index++) };
  }
  prefix(): string {
    return this.input.slice(0, this.index);
  }
  suffix(): string {
    return this.input.slice(this.index);
  }
  [Symbol.iterator]() {
    return this;
  }
}

export function charCodes(input: string) {
  return new CharCodeIterator(input);
}

class ConcatIterator<T> implements Iterator<T> {
  private iters: Iterator<T, T>[];
  private index: number = 0;
  constructor(...iters: Iterator<T>[]) {
    this.iters = iters;
  }
  next(): IteratorResult<T> {
    if (this.index >= this.iters.length) {
      return { done: true, value: undefined };
    }
    const iter = this.iters[this.index];
    const { done, value } = iter.next();
    if (done) {
      this.index++;
      return this.next();
    }
    return { done, value };
  }
}

export function concat<T>(...iters: Iterator<T>[]) {
  return new ConcatIterator(...iters);
}

export class RewindableIterator<T> implements IterableIterator<T> {
  private iter: Iterator<T, T>;
  private buffer: T[] = [];
  private index: number = 0;
  constructor(iter: Iterator<T>) {
    this.iter = iter;
  }
  next(): IteratorResult<T> {
    if (this.index < this.buffer.length) {
      return { done: false, value: this.buffer[this.index++] };
    }
    const { value, done } = this.iter.next();
    if (done) {
      return { done, value: undefined };
    }
    this.buffer.push(value);
    this.index++;
    return { done: false, value };
  }
  reset(n: number) {
    this.buffer = this.buffer.slice(n);
    this.index = 0;
  }
  [Symbol.iterator]() {
    return this;
  }
  get buffered() {
    return this.buffer.length;
  }
}

export class BacktrackableIterator<T> implements IterableIterator<T> {
  private iter: Iterator<T, T>;
  private buffer: T[] = [];
  constructor(iter: Iterator<T>) {
    this.iter = iter;
  }
  [Symbol.iterator]() {
    return this;
  }
  next(): IteratorResult<T> {
    let item = this.buffer.pop();
    if (item !== undefined) {
      return { done: false, value: item };
    }
    return this.iter.next();
  }
  pushBack(item: T) {
    this.buffer.push(item);
  }
}

export function backtrackable<T>(iter: Iterator<T>) {
  return new BacktrackableIterator(iter);
}

export function rewindable<T>(iter: Iterator<T>) {
  return new RewindableIterator(iter);
}

export function iterable<T>(it: Iterator<T>): IterableIterator<T> {
  const iterable = {
    next() {
      return it.next();
    },
    [Symbol.iterator]() {
      return iterable;
    },
  };
  return iterable;
}

export function* range(start: number, end: number) {
  for (let i = start; i < end; i++) {
    yield i;
  }
}
