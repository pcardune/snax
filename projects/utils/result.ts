import { err, Result } from 'neverthrow';

export function flatten<T, E1, E2>(
  result: Result<Result<T, E2>, E1>
): Result<T, E2 | E1> {
  if (result.isOk()) {
    return result.value;
  }
  return err(result.error);
}
