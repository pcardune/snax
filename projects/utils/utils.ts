export function memoize<D>(init: () => D): () => D {
  let cached: D | undefined = undefined;
  return () => {
    if (cached === undefined) {
      cached = init();
    }
    return cached;
  };
}
