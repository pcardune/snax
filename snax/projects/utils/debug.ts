/**
 * Common type declaractions
 */

/**
 * Something that has a debug str
 */
export interface IHaveDebugStr {
  toDebugStr(): string;
}

export function log(...args: any[]) {
  logger.log(...args);
}

type Listener = (...args: any[]) => void;
class Logger {
  static readonly instance = new Logger();
  private listeners: Set<Listener> = new Set();
  private debugFile: any = undefined;

  private constructor() {}

  subscribe(listener: Listener) {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  log(...args: any[]) {
    for (const listener of this.listeners) {
      try {
        listener(...args);
      } catch {}
    }
    if (process.env.DEBUG) {
      console.log(...args);
    } else if (process.env.DEBUG_FILE) {
      const fs = require('fs');
      if (!this.debugFile) {
        this.debugFile = fs.openSync(process.env.DEBUG_FILE, 'w');
      }
      fs.writeSync(this.debugFile, args.join(' ') + '\n');
    }
  }

  capture<R>(insideFunc: () => R, logs: string[]): R {
    const unsub = this.subscribe((...args: any[]) => logs.push(args.join(' ')));
    const result = insideFunc();
    unsub();
    return result;
  }
}
export const logger = Logger.instance;

let shouldUseColors = false;
export function useColors() {
  shouldUseColors = true;
}

const ColorCodes = {
  black: '\u001b[30m',
  red: '\u001b[31m',
  green: '\u001b[32m',
  yellow: '\u001b[33m',
  blue: '\u001b[34m',
  magenta: '\u001b[35m',
  cyan: '\u001b[36m',
  white: '\u001b[37m',
  reset: '\u001b[0m',
  bold: '\u001b[1m',
  underline: '\u001b[4m',
  reversed: '\u001b[7m',
};
type ColorFuncs = {
  [Property in keyof typeof ColorCodes]: (s: string) => string;
};
const colors: ColorFuncs = {} as ColorFuncs;
for (let key in ColorCodes) {
  let color = key as keyof ColorFuncs;
  colors[color] = (s: string): string =>
    shouldUseColors ? ColorCodes[color] + s + ColorCodes.reset : s;
}
export { colors };
