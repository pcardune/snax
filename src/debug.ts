/**
 * Common type declaractions
 */

/**
 * Something that has a debug str
 */
export interface IHaveDebugStr {
  toDebugStr(): string;
}

import fs from 'fs';

let debugFile: any = undefined;
export function log(...args: any[]) {
  if (process.env.DEBUG) {
    console.log(...args);
  } else if (process.env.DEBUG_FILE) {
    if (!debugFile) {
      debugFile = fs.openSync(process.env.DEBUG_FILE, 'w');
    }
    fs.writeSync(debugFile, args.join(' ') + '\n');
  }
}

let shouldUseColors = false;
export function useColors() {
  shouldUseColors = true;
}

const ColorCodes = {
  black: '\u001b[30m',
  red: '\u001b[31m',
  green: '\u001b[32m',
  yellow: '\u001b[33m',
  blue: '\u001b[34m',
  magenta: '\u001b[35m',
  cyan: '\u001b[36m',
  white: '\u001b[37m',
  reset: '\u001b[0m',
  bold: '\u001b[1m',
  underline: '\u001b[4m',
  reversed: '\u001b[7m',
};
type ColorFuncs = {
  [Property in keyof typeof ColorCodes]: (s: string) => string;
};
const colors: ColorFuncs = {} as ColorFuncs;
for (let key in ColorCodes) {
  let color = key as keyof ColorFuncs;
  colors[color] = (s: string): string =>
    shouldUseColors ? ColorCodes[color] + s + ColorCodes.reset : s;
}
export { colors };
