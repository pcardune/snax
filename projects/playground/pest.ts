import { OrderedMap } from '../utils/data-structures/OrderedMap';
import { buildParser } from '../grammar/top-down-parser';
import { charCodes } from '../utils/iter';
import { buildLexer } from '../lexer-gen';
import { parseRegex } from '../regex-compiler/parser';
import { notChars, SingleCharNFA } from '../nfa-to-dfa/regex-nfa';

const re = (s: string) => parseRegex(s).nfa();
const char = (s: string) => new SingleCharNFA(s);

export const lexer = buildLexer(
  new OrderedMap([
    ['ID', re('[a-zA-Z_]([a-zA-Z0-9_]*)')],
    ['=', char('=')],
    ['(', char('(')],
    [')', char(')')],
    ['{', char('{')],
    ['}', char('}')],
    ['~', char('~')],
    ['|', char('|')],
    ['*', char('*')],
    ['+', char('+')],
    ['?', char('?')],
    ['COMMENT', char('/').concat(char('/')).concat(notChars('\n').star())],
    ['STRING', char('"').concat(notChars('"').star()).concat(char('"'))],
    ['WS', re('( |\t)')],
    ['NEWLINE', char('\n')],
  ]),
  ['WS', 'NEWLINE', 'COMMENT']
);

// prettier-ignore
export const parser = buildParser({
  Root: [['Rules']],
  Rules: [
    ['Rule', 'Rules'],
    ['Rule'],
    []
  ],
  Rule: [
    ['ID', '=', '{', 'Expr', '}']
  ],
  Expr: [
    ['Unary', '|', 'Expr'],
    ['Unary', '~', 'Expr'],
    ['Unary']
  ],
  Unary: [
    ['Term', '+'],
    ['Term', '*'],
    ['Term', '?'],
    ['Term']
  ],
  Term: [
    ['(', 'Expr', ')'],
    ['STRING'],
    ['ID']
  ],
});

export class PestParser {
  static parseStr(input: string) {
    return parser.parseTokensOrThrow(lexer.parse(charCodes(input)));
  }
}
