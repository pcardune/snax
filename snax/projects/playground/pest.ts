import { OrderedMap } from '../utils/data-structures/OrderedMap.js';
import { buildParser } from '../grammar/top-down-parser.js';
import { charCodes } from '../utils/iter.js';
import { buildLexer } from '../lexer-gen/index.js';
import { parseRegex } from '../regex-compiler/parser.js';
import { notChars, SingleCharNFA } from '../nfa-to-dfa/regex-nfa.js';

const re = (s: string) => parseRegex(s).nfa();
const char = (s: string) => new SingleCharNFA(s);

export const lexer = buildLexer(
  new OrderedMap([
    ['ID', re('[a-zA-Z_]([a-zA-Z0-9_]*)')],
    ['=', char('=')],
    ['(', char('(')],
    [')', char(')')],
    ['{', char('{')],
    ['}', char('}')],
    ['~', char('~')],
    ['|', char('|')],
    ['*', char('*')],
    ['+', char('+')],
    ['?', char('?')],
    ['COMMENT', char('/').concat(char('/')).concat(notChars('\n').star())],
    ['STRING', char('"').concat(notChars('"').star()).concat(char('"'))],
    ['WS', re('( |\t)')],
    ['NEWLINE', char('\n')],
  ]),
  ['WS', 'NEWLINE', 'COMMENT']
);

// prettier-ignore
export const parser = buildParser({
  Root: [['Rules']],
  Rules: [
    ['Rule', 'Rules'],
    ['Rule'],
    []
  ],
  Rule: [
    ['ID', '=', '{', 'Expr', '}']
  ],
  Expr: [
    ['Unary', '|', 'Expr'],
    ['Unary', '~', 'Expr'],
    ['Unary']
  ],
  Unary: [
    ['Term', '+'],
    ['Term', '*'],
    ['Term', '?'],
    ['Term']
  ],
  Term: [
    ['(', 'Expr', ')'],
    ['STRING'],
    ['ID']
  ],
});

export class PestParser {
  static parseStr(input: string) {
    return parser.parseTokensOrThrow(lexer.parse(charCodes(input)));
  }
}
