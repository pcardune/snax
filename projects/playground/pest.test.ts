import { charCodes } from '../utils/iter';
import { lexer, PestParser } from './pest';
import * as debug from '../utils/debug';
describe('pest parser', () => {
  test('simple', () => {
    const input = 'foo = { "foo" }';
    for (const token of lexer.parse(charCodes(input))) {
      debug.log(token);
    }
    const root = PestParser.parseStr(input);
    if (!root) {
      fail();
    }
    expect(root.pretty()).toMatchSnapshot();
  });
});
