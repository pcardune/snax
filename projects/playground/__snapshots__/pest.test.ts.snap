// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pest parser simple 1`] = `
"
<Root>
|  <Rules>
|  |  <Rule>
|  |  |  <ID>foo</ID>
|  |  |  <=>=</=>
|  |  |  <{>{</{>
|  |  |  <Expr>
|  |  |  |  <Unary>
|  |  |  |  |  <Term>
|  |  |  |  |  |  <STRING>\\"foo\\"</STRING>
|  |  |  |  |  </Term>
|  |  |  |  </Unary>
|  |  |  </Expr>
|  |  |  <}>}</}>
|  |  </Rule>
|  |  <Rules>
|  |  </Rules>
|  </Rules>
</Root>
"
`;
